### PR TITLE
Add basic type check for Item value

### LINKED ISSUE #1517: 能否增加格式校验的hook
## 问题描述
目前apollo只能做一些简单的格式校验，但很多时候，数据本身是带有一定的格式要求的，特别是一些复杂的json、xml配置等等。目前只是对json、xml本身进行了校验，但是没有对内容进行校验。如果内容不符合要求，只有运行时才能发现，无法确定配置是否生效了（可能内容有问题，直接被抛弃了）。如果不及时回滚，甚至会发生服务无法重启之类的问题。

对于这种配置文件，目前的方案是使用open-api，单独搭建一个服务，但这样做略麻烦，而且配置地址会比较分散。

## 建议
增加类似git hook的机制，可以给每个namespace增加一个hook(url 地址)。如果发现有hook，在执行保存前调用一次做一个预校验。还可以增加一个post hook，比如把数据做加密什么的。

### PR DESCRIPTION
**Pull request by furaul**
## What's the purpose of this PR

Add basic type check for Item value

## Which issue(s) this PR fixes:
Fixes #1517

## Brief changelog

Add basic type check for Item value

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Read the [Contributing Guide](https://github.com/ctripcorp/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit tests to verify the code.
- [x] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [x] Update the [`CHANGES` log](https://github.com/ctripcorp/apollo/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by furaul** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Bobji** [comment]
I have read the CLA Document and I hereby sign the CLA

File: apollo-portal/src/main/resources/static/i18n/zh-CN.json [hunk 1]
```diff
@@ -68,15 +68,24 @@
   "Component.ConfigItem.ModifyItem": "修改配置项",
   "Component.ConfigItem.ItemKey": "Key",
   "Component.ConfigItem.ItemValue": "Value",
-  "Component.ConfigItem.ItemValueTips": "注意: 隐藏字符（空格、换行符、制表符Tab）容易导致配置出错，如果需要检测 Value 中隐藏字符，请点击",
-  "Component.ConfigItem.ItemValueShowDetection": "检测隐藏字符",
-  "Component.ConfigItem.ItemValueNotHiddenChars": "无隐藏字符",
+  "Component.ConfigItem.ItemValueTips": "注意: 特殊字符(空格、换行符、制表符Tab、中文逗号)容易导致配置出错，如果需要检测Value中特殊字符，请点击",
```
Comment (mghio) [inline apollo-portal/src/main/resources/static/i18n/zh-CN.json:71]:
```suggestion
  "Component.ConfigItem.ItemValueTips": "注意: 特殊字符(空格、换行符、制表符Tab、中文逗号)容易导致配置出错，如果需要检测 Value 中特殊字符，请点击",
```

**Comment by lepdou** [comment]
add ddl sql to delta file~
image

**Comment by furaul** [comment]
> add ddl sql to delta file~ image

Done!

File: scripts/sql/delta/v190-v200/apolloconfigdb-v190-v200.sql [hunk 2]
```diff
@@ -37,6 +37,7 @@ ALTER TABLE `GrayReleaseRule`
 
 ALTER TABLE `Item`
     ADD COLUMN `DeletedAt` BIGINT(20) NOT NULL DEFAULT '0' COMMENT 'Delete timestamp based on milliseconds' AFTER `IsDeleted`,
+    ADD COLUMN `Type` tinyint(3) unsigned NOT NULL DEFAULT '0' COMMENT '配置项类型，0: String，1: Number，2: Boolean，3: Json' AFTER `Key`,
```
Comment (nobodyiam) [inline scripts/sql/delta/v190-v200/apolloconfigdb-v190-v200.sql:40]: apollo 2.0.0 is released, we need to add the upgrade scripts in a new folder - `v200-v210`.

Comment (furaul) [reply scripts/sql/delta/v190-v200/apolloconfigdb-v190-v200.sql]: TKS. I have already move the ddl to the new file in a new folder - v200-v210

File: apollo-portal/src/main/resources/static/i18n/zh-CN.json [hunk 3]
```diff
@@ -124,6 +133,7 @@
   "Component.Namespace.MasterBranch.Body.PublishState": "发布状态",
   "Component.Namespace.MasterBranch.Body.ItemKey": "Key",
   "Component.Namespace.MasterBranch.Body.ItemValue": "Value",
+  "Component.Namespace.MasterBranch.Body.ItemType": "类型",
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/i18n/zh-CN.json:136]: How about reusing `Component.ConfigItem.ItemTypeName`?

File: apollo-portal/src/main/resources/static/i18n/zh-CN.json [hunk 4]
```diff
@@ -132,6 +142,10 @@
   "Component.Namespace.MasterBranch.Body.ItemNoPublish": "未发布",
   "Component.Namespace.MasterBranch.Body.ItemEffective": "已生效的配置",
   "Component.Namespace.MasterBranch.Body.ItemPublished": "已发布",
+  "Component.Namespace.MasterBranch.Body.ItemTypeString": "String",
+  "Component.Namespace.MasterBranch.Body.ItemTypeNumber": "Number",
+  "Component.Namespace.MasterBranch.Body.ItemTypeBoolean": "Boolean",
+  "Component.Namespace.MasterBranch.Body.ItemTypeJson": "Json",
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/i18n/zh-CN.json:148]: How about reusing `Component.ConfigItem.ItemTypeString`, `Component.ConfigItem.ItemTypeNumber`, etc?

File: apollo-portal/src/main/resources/static/i18n/zh-CN.json [hunk 5]
```diff
@@ -68,15 +68,24 @@
...
   "Component.ConfigItem.ItemJSONValueDetection": "检测JSON格式",
   "Component.ConfigItem.ValidItemJSONValue": "JSON格式正确",
   "Component.ConfigItem.InvalidItemJSONValue": "JSON格式错误",
   "Component.ConfigItem.ItemComment": "Comment",
   "Component.ConfigItem.ChooseCluster": "选择集群",
+  "Component.ConfigItem.ItemTypeName": "类型",
+  "Component.ConfigItem.ItemTypeString": "String",
+  "Component.ConfigItem.ItemTypeNumber": "Number",
+  "Component.ConfigItem.ItemTypeBoolean": "Boolean",
+  "Component.ConfigItem.ItemTypeJson": "Json",
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/i18n/zh-CN.json:84]: How about `JSON`?

**Review by nobodyiam — COMMENTED** [review]
There was a pull request that added JSON check feature #4519. I think this pull request could provide a general type check by combining that PR's codes, e.g. enhance the `check` function that only validates JSON with the ability to check other types.

File: scripts/sql/apolloconfigdb.sql [hunk 6]
```diff
@@ -232,6 +232,7 @@ CREATE TABLE `Item` (
   `Id` int(10) unsigned NOT NULL AUTO_INCREMENT COMMENT '自增Id',
   `NamespaceId` int(10) unsigned NOT NULL DEFAULT '0' COMMENT '集群NamespaceId',
   `Key` varchar(128) NOT NULL DEFAULT 'default' COMMENT '配置项Key',
+  `Type` tinyint(3) unsigned NOT NULL DEFAULT '0' COMMENT '配置项类型，0: String，1: Number，2: Boolean，3: Json',
```
Comment (nobodyiam) [inline scripts/sql/apolloconfigdb.sql:235]:
```suggestion
  `Type` tinyint(3) unsigned NOT NULL DEFAULT '0' COMMENT '配置项类型，0: String，1: Number，2: Boolean，3: JSON',
```

File: scripts/sql/delta/v200-v210/apolloconfigdb-v200-v210.sql [hunk 7]
```diff
@@ -18,4 +18,7 @@
 Use ApolloConfigDB;
 
 -- add INDEX for ReleaseHistory table
-CREATE INDEX IX_PreviousReleaseId ON ReleaseHistory(PreviousReleaseId);
\ No newline at end of file
+CREATE INDEX IX_PreviousReleaseId ON ReleaseHistory(PreviousReleaseId);
+
+ALTER TABLE `Item`
+    ADD COLUMN `Type` tinyint(3) unsigned NOT NULL DEFAULT '0' COMMENT '配置项类型，0: String，1: Number，2: Boolean，3: Json' AFTER `Key`;
```
Comment (nobodyiam) [inline scripts/sql/delta/v200-v210/apolloconfigdb-v200-v210.sql:24]:
```suggestion
    ADD COLUMN `Type` tinyint(3) unsigned NOT NULL DEFAULT '0' COMMENT '配置项类型，0: String，1: Number，2: Boolean，3: JSON' AFTER `Key`;
```

File: apollo-portal/src/main/resources/static/i18n/zh-CN.json [hunk 8]
```diff
@@ -68,15 +68,24 @@
   "Component.ConfigItem.ModifyItem": "修改配置项",
   "Component.ConfigItem.ItemKey": "Key",
   "Component.ConfigItem.ItemValue": "Value",
-  "Component.ConfigItem.ItemValueTips": "注意: 隐藏字符（空格、换行符、制表符Tab）容易导致配置出错，如果需要检测 Value 中隐藏字符，请点击",
-  "Component.ConfigItem.ItemValueShowDetection": "检测隐藏字符",
-  "Component.ConfigItem.ItemValueNotHiddenChars": "无隐藏字符",
+  "Component.ConfigItem.ItemValueTips": "注意: 特殊字符(空格、换行符、制表符Tab、中文逗号)容易导致配置出错，如果需要检测 Value 中特殊字符，请点击",
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/i18n/zh-CN.json:71]: Please also update the [en.json](https://github.com/apolloconfig/apollo/blob/master/apollo-portal/src/main/resources/static/i18n/en.json).

File: apollo-portal/src/main/resources/static/views/component/namespace-panel-master-tab.html [hunk 9]
```diff
@@ -291,7 +291,9 @@
                                     ng-click="col='item.key';desc=!desc;">
                                     {{'Component.Namespace.Master.Items.Body.ItemKey' | translate }}&nbsp;
                                     <span class="glyphicon glyphicon-sort"></span>
-
+                                </th>
+                                <th>
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/views/component/namespace-panel-master-tab.html:295]: Is it better to display the type column after the value column?

**Comment by nobodyiam** [comment]
There are some unit test failures, please also help to fix them.

**Comment by furaul** [comment]
> There was a pull request that added JSON check feature #4519. I think this pull request could provide a general type check by combining that PR's codes, e.g. enhance the `check` function that only validates JSON with the ability to check other types.

I already view #4519, I think it is a good idea to enhance `check` function to remain only one `checkJson` function. There is another question, do I need to remove the json check component in #4519, to remain only one 'json check' component?

**Comment by AbnerHuang2** [comment]
if in this way, don't forget openapi module.  openapi may also need to be compatible with type.

**Comment by nobodyiam** [comment]
@furaul 

> There is another question, do I need to remove the json check component in https://github.com/apolloconfig/apollo/pull/4519, to remain only one 'json check' component?

I think we could merge and leave one `json check` component.

**Comment by nobodyiam** [comment]
> if in this way, don't forget openapi module. openapi may also need to be compatible with type.

This is a good point.

File: apollo-portal/src/main/resources/static/i18n/zh-CN.json (hunk 8 above)
Comment (furaul) [reply apollo-portal/src/main/resources/static/i18n/zh-CN.json:71]: > Please also update the [en.json](https://github.com/apolloconfig/apollo/blob/master/apollo-portal/src/main/resources/static/i18n/en.json).

Fixed it.

File: apollo-portal/src/main/resources/static/views/component/namespace-panel-master-tab.html [hunk 10]
```diff
@@ -32,8 +32,8 @@
             <div class="col-md-6 col-sm-6 text-right header-buttons">
 
                 <button type="button" class="btn btn-default btn-sm" data-tooltip="tooltip" data-placement="bottom"
-                    title="{{'Component.Namespace.Master.LoadNamespaceTips' | translate }}"
-                    ng-click="refreshNamespace()">
+                        title="{{'Component.Namespace.Master.LoadNamespaceTips' | translate }}"
+                        ng-click="refreshNamespace()">
```
Comment (furaul) [inline apollo-portal/src/main/resources/static/views/component/namespace-panel-master-tab.html:36]: I do not know why there are some diffs, in my PC local, the line is not changed.

**Comment by furaul** [comment]
> There are some unit test failures, please also help to fix them.

Fixed them, please trigger the unit tests to run again.

**Comment by furaul** [comment]
> @furaul
> 
> > There is another question, do I need to remove the json check component in #4519, to remain only one 'json check' component?
> 
> I think we could merge and leave one `json check` component.

@nobodyiam @AbnerHuang2 Finished, and openapi module is processed. Please review it.

Discussed with nobodyiam on WeChat, only add type and basic check in openapi module, and it is compatiable with old version SDKs.

**Comment by nobodyiam** [comment]
This looks quite good!
BTW, shall we also add the item type feature to the grayscale namespace and associated namespace?

image

image

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/server/service/ServerItemOpenApiService.java [hunk 11]
```diff
@@ -72,6 +72,7 @@ public void updateItem(String appId, String env, String clusterName, String name
         .loadItem(Env.valueOf(env), appId, clusterName, namespaceName, itemDTO.getKey());
     //protect. only value,comment,lastModifiedBy can be modified
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/openapi/server/service/ServerItemOpenApiService.java:73]:
```suggestion
    //protect. only value,type,comment,lastModifiedBy can be modified
```

File: apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/ItemController.java [hunk 12]
```diff
@@ -143,7 +143,8 @@ public ItemDTO update(@PathVariable("appId") String appId,
 
     Item beforeUpdateItem = BeanUtils.transform(Item.class, managedEntity);
 
-    //protect. only value,comment,lastModifiedBy can be modified
+    //protect. only value,type,comment,lastModifiedBy can be modified
+    managedEntity.setType(entity.getType());
```
Comment (nobodyiam) [inline apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/controller/ItemController.java:147]: I think `com.ctrip.framework.apollo.biz.service.ItemSetService#doUpdateItems` should also be updated.

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**